src/ZoneServer/World/Storage/Storage.cs

```diff
@@ -0,0 +1,691 @@
+		/// If item is not stackable, returns 0.
+		/// </summary>
+		/// <param name="item">Item to add</param>
+		/// <param name="amount">Amount to stack to item</param>
+		/// <returns>Amount of stack added</returns>
+		private int AddItemStack(Item item, int amount)
+		{
+			if (!item.IsStackable)
+				return 0;
+
+			if (amount <= 0)
+				return 0;
+
+			if ((item.Amount + amount) < item.Data.MaxStack)
+			{
+				item.Amount += amount;
+				return amount;
+			}
+			else
+			{
+				var amountToAdd = item.Data.MaxStack - item.Amount;
+				item.Amount += amountToAdd;
+				return amountToAdd;
+			}
+		}
+
+		/// <summary>
+		/// Looks in storage for items that the given item
+		/// can stack to. Does nothing if given item is not stackable.
+		/// </summary>
+		/// <param name="item">Item type look for</param>
+		/// <param name="foundStorageItems">Position and Item object of found items</param>
+		private void TryGetStackableItems(Item item, out SortedList<int, Item> foundStorageItems)
+		{
+			foundStorageItems = new SortedList<int, Item>();
+
+			// Stacks to existing items
+			if (item.IsStackable)
+			{
+				lock (_syncLock)
+				{
+					foreach (var storageItem in _storageItems)
+					{
+						var existingItem = storageItem.Value;
+						var existingItemPosition = storageItem.Key;
+						if (existingItem.IsStackable && (existingItem.Data.ClassName == item.Data.ClassName))
+						{
+							foundStorageItems.Add(existingItemPosition, existingItem);
+						}
+					}
+				}
+			}
+		}
+
+		/// <summary>
+		/// Calculates the amount of given item that can be added to storage.
+		/// </summary>
+		/// <param name="item">Item and the amount to check</param>
+		/// <returns></returns>
+		private int GetItemAvailableAmount(Item item)
+		{
+			int availableAmount = 0;
+
+			// Checks if stackable item can be placed in storage
+			if (item.IsStackable)
+			{
+				this.TryGetStackableItems(item, out var existingStorageItems);
+				var stackAvailableAmount = existingStorageItems.Values.Sum(existingItem => existingItem.Data.MaxStack - existingItem.Amount);
+
+				// No space in existing stacks and no new position available
+				if (this.CheckStorageFull() && stackAvailableAmount <= 0)
+					return 0;
+
+				// Existing stacks can't stack and adding to new position is disabled
+				if (!ZoneServer.Instance.Conf.World.StorageMultiStack && stackAvailableAmount <= 0)
+					return 0;
+
+				availableAmount = stackAvailableAmount;
+
+				// We can also add the stack to empty spaces if multistack is enabled
+				if (ZoneServer.Instance.Conf.World.StorageMultiStack && !this.CheckStorageFull())
+					availableAmount = item.Data.MaxStack;
+			}
+			else
+			{
+				// Checks if non stackable item can be placed in storage
+				if (this.CheckStorageFull())
+					return 0;
+
+				availableAmount = item.Amount;
+			}
+
+			return availableAmount;
+		}
+
+		/// <summary>
+		/// Adds the available amount of an item to storage.
```

Comment:
Unfortunately this comment and the name of the method aren't very descriptive. What is the "available amount", what happens to the item, what are the potential results? You essentially have to read the method to understand what it does.